### PR TITLE
FEAT-#6970: Implement to/from_ray_dataset functions

### LINKED ISSUE #6970: Implement to/from_ray_dataset functions

### PR DESCRIPTION
**Pull request by Retribution98**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6970 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/test/test_io.py [hunk 1]
```diff
@@ -3228,3 +3228,44 @@ def test_to_period():
...
+
+    if Engine.get() != "Ray":
+        ray_df = modin_df.to_ray()
+        df_equals(ray_df.to_pandas(), pandas_df)
+    else:
+        with pytest.raises(
+            RuntimeError,
+            match="Modin Dataframe may only be converted to a Ray Dataset if Modin uses a Ray engine.",
+        ):
+            ray_df = modin_df.to_ray()
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:3249]:
```suggestion
    if Engine.get() == "Ray":
        ray_df = modin_df.to_ray()
        df_equals(ray_df.to_pandas(), pandas_df)
    else:
        with pytest.raises(
            RuntimeError,
            match="Modin Dataframe may only be converted to a Ray Dataset if Modin uses a Ray engine.",
        ):
            _= modin_df.to_ray()
```

File: modin/pandas/test/test_io.py [hunk 2]
```diff
@@ -3228,3 +3228,44 @@ def test_to_period():
...
+
+    if Engine.get() != "Ray":
+        result_df = from_ray(ray_df)
+        df_equals(result_df, modin_df)
+    else:
+        with pytest.raises(
+            RuntimeError,
+            match="Modin Dataframe may only be converted to a Ray Dataset if Modin uses a Ray engine.",
+        ):
+            result_df = from_ray(ray_df)
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:3271]:
```suggestion
    if Engine.get() == "Ray":
        result_df = from_ray(ray_df)
        df_equals(result_df, modin_df)
    else:
        with pytest.raises(
            RuntimeError,
            match="Modin Dataframe may only be converted to a Ray Dataset if Modin uses a Ray engine.",
        ):
            _ = from_ray(ray_df)
```

File: modin/core/execution/ray/generic/io/io.py [hunk 3]
```diff
@@ -18,3 +18,15 @@
 
 class RayIO(BaseIO):
     """Base class for doing I/O operations over Ray."""
+
+    @classmethod
+    def from_ray(cls, ray_obj):
+        raise NotImplementedError(
+            f"Modin dataset cann't be created from `ray.data.Dataset` using {cls}."
+        )
```
Comment (anmyachev) [inline modin/core/execution/ray/generic/io/io.py:26]: Why? Is the PR in draft?

Comment (Retribution98) [reply modin/core/execution/ray/generic/io/io.py:26]: The default implementation (in the BaseIO class) throws an exception stating that _the Ray Engine is required for the `to_ray`/`from_ray` functions_. We also have an additional base class io for the Ray engine that will be used for every data structure implementation using the Ray engine, so we expect all its subclasses to implement these functions in their own way. If we create a new subclass of RayIO and forget to implement these functions, we will get a Not Implemented exception.

In this PR I hasn't added the ``to_ray`/`from_ray` implementation to ``cudf_on_ray`` because it works incorrectly if I'm not mistaken. If we try to use it and call `to_ray`/`from_ray` functions, we will also receive a Not Implemented exception.

File: docs/ecosystem.rst [hunk 4]
```diff
@@ -45,5 +45,19 @@ where NumPy can be used and what libraries it powers.
...
+
+.. code-block:: python
+
+    from modin.pandas.io import to_ray
+
+    pandas_df = to_ray(modin_df)
+
 .. _pandas ecosystem: https://pandas.pydata.org/community/ecosystem.html
 .. _NumPy ecosystem: https://numpy.org
+.. _Ray ecosystem: https://docs.ray.io/en/latest/data/data.html
```
Comment (YarShev) [inline docs/ecosystem.rst:62]:
```suggestion
.. _`Ray Data`: https://docs.ray.io/en/latest/data/data.html
```

File: docs/ecosystem.rst [hunk 5]
```diff
@@ -45,5 +45,19 @@ where NumPy can be used and what libraries it powers.
 
     numpy_arr = to_numpy(modin_df)
 
+to_ray
+---------
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply docs/ecosystem.rst:49]: Why?

File: modin/pandas/test/test_io.py [hunk 6]
```diff
@@ -3228,3 +3228,39 @@ def test_to_period():
...
+
+    modin_df, pandas_df = create_test_dfs(TEST_DATA, index=index)
+
+    if Engine.get() == "Ray":
+        ray_df = modin_df.to_ray()
+        df_equals(ray_df.to_pandas(), pandas_df)
+    else:
+        with pytest.raises(
+            RuntimeError,
+            match="Modin Dataframe may only be converted to a Ray Dataset if Modin uses a Ray engine.",
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply modin/pandas/test/test_io.py]: Done

File: modin/core/execution/ray/generic/io/io.py [hunk 7]
```diff
@@ -18,3 +18,51 @@
...
+        ray.data.Dataset
+            Converted object with type depending on input.
+
+        Notes
+        -----
+        This function must be implemented in every subclass 
+        otherwise NotImplementedError will be raised.
+        """
+        raise NotImplementedError(
+            f"`ray.data.Dataset` cann't be created from Modin DataFrame using {cls}."
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply modin/core/execution/ray/generic/io/io.py]: Done

File: modin/core/execution/ray/generic/io/io.py [hunk 8]
```diff
@@ -18,3 +18,51 @@
...
+        BaseQueryCompiler
+            QueryCompiler containing data from the Ray Dataset.
+
+        Notes
+        -----
+        This function must be implemented in every subclass 
+        otherwise NotImplementedError will be raised.
+        """
+        raise NotImplementedError(
+            f"Modin dataset cann't be created from `ray.data.Dataset` using {cls}."
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply modin/core/execution/ray/generic/io/io.py]: Done

File: docs/ecosystem.rst (hunk 4 above)
Comment (Retribution98) [reply docs/ecosystem.rst]: Done

File: docs/ecosystem.rst [hunk 9]
```diff
@@ -45,5 +45,19 @@ where NumPy can be used and what libraries it powers.
...
+
+You can refer to `Ray ecosystem`_ page to get more details on
+where Ray Dataset can be used and what libraries it powers.
+
+.. code-block:: python
+
+    from modin.pandas.io import to_ray
+
+    pandas_df = to_ray(modin_df)
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply docs/ecosystem.rst]: Done

File: docs/ecosystem.rst [hunk 10]
```diff
@@ -45,5 +45,19 @@ where NumPy can be used and what libraries it powers.
 
     numpy_arr = to_numpy(modin_df)
 
+to_ray
+---------
+
+You can refer to `Ray ecosystem`_ page to get more details on
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply docs/ecosystem.rst]: Done

File: docs/ecosystem.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (Retribution98) [reply docs/ecosystem.rst]: Done

File: modin/pandas/io.py [hunk 11]
```diff
@@ -1034,6 +1034,56 @@ def from_dataframe(df):
...
+    Ray Dataset can only be converted to Modin Dataframe if Modin uses a Ray engine.
+    """
+    from modin.core.execution.dispatching.factories.dispatcher import FactoryDispatcher
+
+    return ModinObjects.DataFrame(
+        query_compiler=FactoryDispatcher.from_ray_dataset(ray_obj)
+    )
+
+
+def from_dask(ray_obj):
```
Comment (YarShev) [inline modin/pandas/io.py:1062]: This should be in a separate PR.

**Review by YarShev — COMMENTED** [review]
Left a comment. Ohwerwise, looks good!

File: modin/pandas/io.py (hunk 11 above)
Comment (Retribution98) [reply modin/pandas/io.py:1062]: Oops, sorry

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py [hunk 12]
```diff
@@ -39,6 +39,10 @@
     PandasSQLParser,
 )
 from modin.core.storage_formats.pandas.query_compiler import PandasQueryCompiler
+from modin.distributed.dataframe.pandas.partitions import (
+    from_partitions,
+    unwrap_partitions,
+)
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:45]: Might it be the case with cycle import? What if we move these imports into from\_/to\_ray_dataset functions?

File: modin/pandas/dataframe.py [hunk 13]
```diff
@@ -44,7 +44,7 @@
 from modin.error_message import ErrorMessage
 from modin.logging import disable_logging
 from modin.pandas import Categorical
-from modin.pandas.io import from_non_pandas, from_pandas, to_pandas
+from modin.pandas.io import from_non_pandas, from_pandas, to_pandas, to_ray_dataset
```
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/pandas/dataframe.py:47]: Look at this comment https://github.com/modin-project/modin/pull/6971/files#r1507763590.

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py (hunk 12 above)
Comment (Retribution98) [reply modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:45]: I can try, but I think we need a refactoring to move the `unwrap_partitions` and `from_partitions` implementations deeper.

File: modin/pandas/dataframe.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/pandas/dataframe.py:47]: This one

> Might it be the case with cycle import? What if we move these imports into from_/to_ray_dataset functions?

**Comment by YarShev** [comment]
@Retribution98, CI fails for the new functions? Can you take a look?

**Comment by YarShev** [comment]
It looks like a Ray Dataset issue. We can submit this bug to them. To workaround the issue for now, I guess we could use a simpler data without a Datetime index for testing.

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py [hunk 14]
```diff
@@ -258,3 +258,45 @@ def func(df, **kw):  # pragma: no cover
...
+        ----------
+        modin_obj : modin.DataFrame
+            The Modin DataFrame to convert.
+
+        Returns
+        -------
+        ray.data.Dataset
+            Converted object with type depending on input.
+        """
+        from ray.data import from_pandas_refs
```
Comment (YarShev) [inline modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:297]: This import could be at the top.

File: modin/pandas/dataframe.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/pandas/dataframe.py:47]: It looks like the cycle import has gone away after moving imports.

**Comment by YarShev** [comment]
@anmyachev, any other comments?

**Comment by Retribution98** [comment]
> It looks like a Ray Dataset issue. We can submit this bug to them. To workaround the issue for now, I guess we could use a simpler data without a Datetime index for testing.

Yes, I know about this problem because I caught it when I was testing this feature locally. This problem is not related to the use of the index, but to the use of an old version of Python from Ray, so I'm very surprised that the previous CI running passed.
In any case, this problem has already been fixed in the master branch of the Ray repository, so we can try installing ray from the nightly releases. I checked it works.

File: docs/ecosystem.rst [hunk 15]
```diff
@@ -45,5 +45,19 @@ where NumPy can be used and what libraries it powers.
...
+--------------
+
+You can refer to `Ray Data`_ page to get more details on
+where Ray Dataset can be used and what libraries it powers.
+
+.. code-block:: python
+
+    from modin.pandas.io import to_ray_dataset
+
+    ray_dataset = to_ray_dataset(modin_df)
```
Comment (dchigarev) [inline docs/ecosystem.rst:58]: maybe we should promote an easier way here?
```suggestion

    ray_dataset = modin_df.to_ray_dataset()
```

**Comment by YarShev** [comment]
Could you point me out to the issue they resolved? I am kind of confused that it relates to the python version Ray uses.

File: docs/ecosystem.rst (hunk 15 above)
Comment (YarShev) [reply docs/ecosystem.rst:58]: I think we could do that in a separate PR for all methods in one go.

File: modin/pandas/dataframe.py [hunk 16]
```diff
@@ -2229,6 +2229,21 @@ def to_parquet(
             **kwargs,
         )
 
+    def to_ray_dataset(self):
```
Comment (anmyachev) [inline modin/pandas/dataframe.py:2232]: What about Series?

Comment (YarShev) [reply modin/pandas/dataframe.py:2232]: I think it is okay to have a one column DataFrame out of a one column Ray Dataset. If the user needs a series, just call .squeeze(axis=1).

**Comment by Retribution98** [comment]
> Could you point me out to the issue they resolved? I am kind of confused that it relates to the python version Ray uses.

This problem was resolved [here](https://github.com/ray-project/ray/pull/42892).

File: modin/pandas/dataframe.py (hunk 16 above)
Comment (Retribution98) [reply modin/pandas/dataframe.py:2232]: > Many pandas operations return different types depending on the data, so we force users to check the type of the result of any operation every time.

I missed understanding. What Ray type do you suggest converting Series instead of Dataset?

Comment (anmyachev) [reply modin/pandas/dataframe.py:2232]: `modin.pandas.Series` doesn't have `to_ray_dataset` but `modin.pandas.DataFrame` has. In order not to receive errors such as the absence of an attribute on an object, the user will have to check the type.

Comment (Retribution98) [reply modin/pandas/dataframe.py:2232]: Ok, thanks for your clarification. I think it will be really more convenient if we add `to_ray_dataset` function to `modin.pandas.Series`

**Comment by YarShev** [comment]
It seems we should xfail the test until a new release of Ray is off and then enable it.

**Comment by Retribution98** [comment]
> I see, it was related to a new pandas but not python. Thanks!

The pandas version, of course, sorry for this misprint.

File: modin/pandas/series.py [hunk 17]
```diff
@@ -32,7 +32,7 @@
 
 from modin.config import PersistentPickle
 from modin.logging import disable_logging
-from modin.pandas.io import from_pandas, to_pandas
+from modin.pandas.io import from_pandas, to_pandas, to_ray_dataset
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply modin/pandas/series.py:35]: @YarShev
Apparently that change still doesn’t work.
It looks like a false positive, CodeQL have the same [PR](https://github.com/github/codeql/issues/13866).

File: modin/pandas/dataframe.py (hunk 16 above)
Comment (Retribution98) [reply modin/pandas/dataframe.py:2232]: Done

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py (hunk 14 above)
Comment (Retribution98) [reply modin/core/execution/ray/implementations/pandas_on_ray/io/io.py]: Yes, it could be. Done

File: modin/pandas/series.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/pandas/series.py:35]: Oh, I see. Then we can revert imports back to the top.

File: modin/pandas/test/test_io.py [hunk 18]
```diff
@@ -3228,3 +3228,72 @@ def test_to_period():
...
+        df_equals(ray_dataset.to_pandas(), pandas_df)
+    else:
+        with pytest.raises(
+            RuntimeError,
+            match="Modin Dataframe can only be converted to a Ray Dataset if Modin uses a Ray engine.",
+        ):
+            _ = modin_df.to_ray_dataset()
+
+
+@pytest.mark.xfail(
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:3256]: same

File: modin/pandas/test/test_io.py [hunk 19]
```diff
@@ -3228,3 +3228,72 @@ def test_to_period():
...
+)
+@pytest.mark.filterwarnings(default_to_pandas_ignore_string)
+def test_from_ray_dataset():
+    index = pandas.DatetimeIndex(
+        pandas.date_range("2000", freq="h", periods=len(TEST_DATA["col1"]))
+    )
+    modin_df, pandas_df = create_test_dfs(TEST_DATA, index=index)
+
+    ray_df = ray.data.from_pandas(pandas_df)
+
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:3297]:
```suggestion
```

File: modin/pandas/test/test_io.py [hunk 20]
```diff
@@ -3228,3 +3228,72 @@ def test_to_period():
     )
     modin_df, pandas_df = create_test_dfs(TEST_DATA, index=index)
     df_equals(modin_df.to_period(), pandas_df.to_period())
+
+
+@pytest.mark.xfail(
+    Engine.get() == "Ray",
+    reason="Ray-2.9.3 has a problem using pandas 2.2.0. It will be resolved in the next release of Ray.",
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply modin/pandas/test/test_io.py:3235]: Done

File: modin/pandas/test/test_io.py (hunk 18 above)
Comment (Retribution98) [reply modin/pandas/test/test_io.py:3256]: Done

File: modin/pandas/test/test_io.py (hunk 19 above)
Comment (Retribution98) [reply modin/pandas/test/test_io.py]: Done

File: modin/pandas/test/test_io.py [hunk 21]
```diff
@@ -3228,3 +3228,71 @@ def test_to_period():
...
+@pytest.mark.xfail(
+    Engine.get() == "Ray" and version.parse(ray.__version__) <= version.parse("2.9.3"),
+    reason="Ray-2.9.3 has a problem using pandas 2.2.0. It will be resolved in the next release of Ray.",
+)
+@pytest.mark.filterwarnings(default_to_pandas_ignore_string)
+def test_df_to_ray_dataset():
+    index = pandas.DatetimeIndex(
+        pandas.date_range("2000", freq="h", periods=len(TEST_DATA["col1"]))
+    )
+
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:3242]:
```suggestion
```

File: modin/pandas/test/test_io.py [hunk 22]
```diff
@@ -3228,3 +3228,71 @@ def test_to_period():
...
+            match="Modin Dataframe can only be converted to a Ray Dataset if Modin uses a Ray engine.",
+        ):
+            _ = modin_s.to_ray_dataset()
+
+
+@pytest.mark.xfail(
+    Engine.get() == "Ray" and version.parse(ray.__version__) <= version.parse("2.9.3"),
+    reason="Ray-2.9.3 has a problem using pandas 2.2.0. It will be resolved in the next release of Ray.",
+)
+@pytest.mark.skipif(
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:3285]: Why do we skip a test for `from_ray_dataset` for non Ray engines but keep tests for `to_ray_dataset` for those engines? Let's be consistent, i.e., either skip for both methods or run tests for them.

File: modin/core/execution/ray/generic/io/io.py [hunk 23]
```diff
@@ -18,3 +18,51 @@
...
+        )
+
+    @classmethod
+    def to_ray_dataset(cls, modin_obj):
+        """
+        Convert a Modin DataFrame/Series to a Ray Dataset.
+
+        Parameters
+        ----------
+        modin_obj : modin.DataFrame, modin.Series
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:53]:
```suggestion
        modin_obj : modin.pandas.DataFrame, modin.pandas.Series
```

File: modin/core/io/io.py [hunk 24]
```diff
@@ -694,3 +718,27 @@ def to_parquet(cls, obj, path, **kwargs):  # noqa: PR01
...
         return obj.to_parquet(path, **kwargs)
+
+    @classmethod
+    def to_ray_dataset(cls, modin_obj):
+        """
+        Convert a Modin DataFrame/Series to a Ray Dataset.
+
+        Parameters
+        ----------
+        modin_obj : modin.DataFrame, modin.Series
```
Comment (YarShev) [inline modin/core/io/io.py:729]:
```suggestion
        modin_obj : modin.pandas.DataFrame, modin.pandas.Series
```

File: modin/pandas/test/test_io.py (hunk 22 above)
Comment (Retribution98) [reply modin/pandas/test/test_io.py:3284]: I can do it, if you think it is better.

**Review by dchigarev — APPROVED** [review]
